Implemented question bank

dmols opened the pull request:
Question bank is functional and has similar styling as the WIP QB for the original Matching widget. Please test it when you get a chance and let me know if anything breaks or if I should change anything!

dmols committed:
Implemented question bank and added styling

cayb0rg reviewed (changes requested):
Cool feature! This looks good, and most of what I found is just nit-picking.
* I like the short-circuit evaluation made when setting the question bank val; this could maybe be extended with an `||` to produce some sort of error message when the user enters an invalid amount. Adding a `min` attribute on the input would also prevent users from inputting values < 1.
* I think the question bank button could be a little wider so it's all on one line, and I think the blue outline could be confused for focusing. A box shadow instead of a border might match the other buttons better.

dmols committed:
added on-hover CSS and element centering + watch function for questionBankVal

dmols commented:
Thank you for the feedback! I initially tried to use an error message to let the user know to pick a valid input, but it got too convoluted, especially since they could still close the dialog with an incorrect value. Instead, I opted to just make a `watch` function that tracks `questionBankVal` and doesn't allow it to go lower than 1 (unless there are no pairs) or higher than the number of pairs. I also went ahead and made the question bank button bigger and added a box shadow. Lastly, I incorporated some more suggestions seen in the matching widget PR (such as centering the elements and adding on-hover effects). 

Please let me know if it could use anything else. Thank you!

cayb0rg reviewed (approved):
The changes look great! Apart from maybe removing the standard border from the two buttons, I think this looks good to go. 

Also, I can't type in the number of questions to ask without it going back to '1' unless I select and replace, but that's probably going to be inevitable given Angular wackiness.

dmols committed:
Added more css for dialog and replaced questionBankVal listener with disabled attribute solution

dmols committed:
Updated test snapshots

dmols committed:
Added edge case protection for qb shuffle, test coverage, and lowered test threshold

dmols committed:
Changed dialog button styling to fit the theme better

clpetersonucf commented on src/controllers/player.coffee:
```diff
@@ -481,7 +489,7 @@ Matching.controller 'matchingPlayerCtrl', ['$scope', '$timeout', '$sce', ($scope
 		Materia.Engine.end true
 
-	_shuffle = (qsetItems) ->
+	$scope._shuffle = (qsetItems) ->
```
I'm not _crazy_ about the shuffle function being attached to scope, since it appears that was only done to facilitate testing. My first reaction is the manner in which you're conducting the test should change so you don't have to directly reference the function (ex: compare the initial qset coming from `qset.data` with the qset that's provided after the `start` method is called).

However, if shuffle has to remain attached to scope, you should drop the prepended underscore. We usually define functions as `_function` to indicate private methods: ones that are only referenced internally. `$scope._shuffle` should just be `$scope.shuffle`. You'll have to update every other reference accordingly.

dmols replied:
Good point. I opted for reverting the naming for the function, and deleted the previous test function since it wasn't going to work without `_shuffle` being public (not as is anyways). I believe I can't go with the first idea since it's possible the updated array of questions/answers has the exact same order as the initial one, so I need some way to tell if the function is called. 

Coverage is at passing level without code to test that function. That being said, it seems like there's a way to pull in this function while keeping it private without attaching it to scope, but I wanted to bring that up to hear your input first.

clpetersonucf replied:
You're right, it's not really reasonable to test a shuffle function by comparing the order of items before and after - because it's random. I think it would be sufficient to compare the qset item count before and after, and ensure the function was called when the conditions for it to be triggered have been met (question bank is enabled or the randomize option is enabled, etc). If the function can be referenced in tests without attaching it to scope, that's all the better.

clpetersonucf commented on src/creator.html:
```diff
@@ -20,12 +20,37 @@
 	</head>
 	<body ng-app='matching' ng-controller="matchingCreatorCtrl">
 		<div id='modal'></div>
+		<div ng-show= "questionBankDialog" class="overlay"></div>
```
You should probably reuse the existing `modalbg` instead of inventing a second one just for the question bank dialog.

clpetersonucf commented on src/creator.html:
```diff
@@ -20,12 +20,37 @@
+		<div class="questionBankDialog-wrapper" ng-show= "questionBankDialog">
+			<div class="questionBankDialog">
+				<div class="enable-qb-question">
+					<strong>Enable question bank? </strong>
```
Use semantic HTML here, not `<strong>`.

clpetersonucf commented on src/creator.html:
```diff
@@ -20,12 +20,37 @@
+				</div>
+				<div>
+					<span ng-show="enableQuestionBank" >
+						<strong>How many questions to ask? </strong>
```
Same as line 34.

clpetersonucf reviewed (changes requested):
Since your changes here are so similar to those made in This or That, a lot of the same feedback applies:

In your CSS:
- Use previously defined sass variables (`$variable`) or hex values (`#ffffff`) for color values, not words (`black`).
- Selectors (and associated classes/ids) should be `lower-case-separated-by-hyphens`.

In your JS:
- The initial quantity of the question bank, when enabled, should match the current question count, not just `1`.
- There's an edge case where you can select the question bank dialog while the question count is still 0 (you haven't added any word pairs yet). Doing so prevents you from exiting the dialog, because it requires a question bank value > 0. The question bank button in the header should probably be disabled until at least 1 word pair is present.
- If you reuse the modal bg (which you should), its functionality should be extended to also dismiss the question bank dialog when clicked. This always gives users a way out without having to click the confirm option.

clpetersonucf commented:
Some additional style suggestions, if you want to give it a shot:
- The question bank button in the header should visually change based on whether or not the feature is enabled (not just color, but text content).
- "Action buttons" in the creator are rectangular (like the " + Add Pair" button) and flat (no bezel), so I would stick to that visual language.
- The QB button in the header should be right-aligned.
- I actually really like the purple color you went with, and would recommend staying with it and using it as the "action button" color instead of the gray used in the intro dialog. Use it for the confirm dialog within the QB modal as well.

dmols committed:
Fixed color reference names

dmols committed:
Swapped css selector names to snake-case and added extra condition to disable attribute for qb confirm

dmols committed:
Added placeholder for questionBankVal, input validation, and fixed html syntax

dmols committed:
Added small updates to functionality and all requested style updates

dmols committed:
Reverted changes to shuffle function name and testing

dmols committed:
Added test coverage for enableQuestionBank case

dmols committed:
cleaned up commented out css code

dmols committed:
very minor css change for disabled qb button

dmols committed:
Added style changes to qb button and dialog close button

dmols committed:
disabled gray-out when no word pairs exist

dmols committed:
Simplified questionBankVal validation and added test coverage for function

dmols committed:
Minor styling changes

dmols commented:
Final changes have been pushed. Should be good to review!

clpetersonucf commented on src/controllers/creator.coffee:
```diff
@@ -27,6 +31,9 @@ Matching.controller 'matchingCreatorCtrl', ['$scope', '$sce', ($scope, $sce) ->
 
 	materiaCallbacks.initExistingWidget = (title, widget, qset, version, baseUrl) ->
 		_items = qset.items[0].items
+		$scope.enableQuestionBank = if qset.options.enableQuestionBank then qset.options.enableQuestionBank else false
```
Note that while the demo has a top-level `options` object, most LCC qsets don't appear to. You should ensure `qset.options` exists prior to and in addition to checking for the values within `qset.options`

clpetersonucf replied:
This applies to the other `qset.options` references directly below this line as well.

clpetersonucf commented on src/controllers/player.coffee:
```diff
@@ -45,6 +45,14 @@ Matching.controller 'matchingPlayerCtrl', ['$scope', '$timeout', '$sce', ($scope
 		$scope.title = instance.name
+
+		# Update qset items to only include the number of questions specified in the question bank. Done here since $scope.totalItems depends on it.
+		if qset.options.enableQuestionBank
```
Same feedback as above - if `qset` does not have an `options` property, this will throw an error.

clpetersonucf reviewed (changes requested):
Only one final thing I noticed related to references to the `options` property of `qset` which may not exist. When those references are fixed, this is good to go.

Side note: since we're combining this feature with Dee's accessibility updates, I went ahead and created a `dev/2.1.0` branch to stage the changes before they reach master. I updated the base branch target of this PR to point to it instead of master.

dmols committed:
Added safeguarding for older qsets

clpetersonucf reviewed (approved):
Nice work. LGTM.
